ir/src/error.rs

```diff
@@ -74,4 +80,150 @@ impl SemanticError {
+        ))
+    }
+
+    // --- DECLARATION ERRORS ---------------------------------------------------------------------
+
+    fn missing_declaration(missing_section: &str) -> Self {
+        SemanticError::MissingDeclaration(format!("{missing_section} section is missing"))
+    }
+
+    pub(super) fn missing_trace_columns_declaration() -> Self {
+        Self::missing_declaration("trace_declaration")
+    }
+
+    pub(super) fn missing_public_inputs_declaration() -> Self {
+        Self::missing_declaration("public_inputs")
+    }
+
+    pub(super) fn missing_boundary_constraints_declaration() -> Self {
+        Self::missing_declaration("boundary_constraints")
+    }
+
+    pub(super) fn missing_integrity_constraints_declaration() -> Self {
+        Self::missing_declaration("integrity_constraints")
+    }
+
+    pub(super) fn has_random_values_but_missing_aux_trace_columns_declaration() -> Self {
+        SemanticError::MissingDeclaration(
+            "random_values section requires aux_trace_columns section, which is missing"
+                .to_string(),
+        )
+    }
+
+    pub(super) fn illegal_identifer_type(ident_name: &str, ident_type: &IdentifierType) -> Self {
+        SemanticError::InvalidUsage(format!(
+            "Identifier {ident_name} was declared as a {ident_type} which is not a supported type."
+        ))
+    }
+
+    // --- ILLEGAL IDENTIFIER ERRORS --------------------------------------------------------------
+
+    pub(super) fn duplicate_identifer(
+        ident_name: &str,
+        ident_type: IdentifierType,
+        prev_type: IdentifierType,
+    ) -> Self {
+        SemanticError::DuplicateIdentifier(format!(
+            "Cannot declare {ident_name} as a {ident_type}, since it was already defined as a {prev_type}"))
+    }
+
+    pub(super) fn undeclared_identifier(ident_name: &str) -> Self {
+        SemanticError::InvalidIdentifier(format!("Identifier {ident_name} was not declared"))
+    }
+
+    pub(super) fn illegal_trace_column_identifier(
+        ident_name: &str,
+        ident_type: &IdentifierType,
+    ) -> Self {
+        SemanticError::InvalidUsage(format!(
+            "Identifier {ident_name} was declared as a {ident_type} not as a trace column"
+        ))
+    }
+
+    // --- ILLEGAL VALUE ERRORS -------------------------------------------------------------------
+
+    pub(super) fn cycle_length_not_power_of_two(length: usize, cycle_name: &str) -> Self {
+        SemanticError::InvalidPeriodicColumn(format!(
+            "cycle length must be a power of two, but was {length} for cycle {cycle_name}"
+        ))
+    }
+
+    pub(super) fn cycle_length_too_small(length: usize, cycle_name: &str) -> Self {
+        SemanticError::InvalidPeriodicColumn(format!(
+            "cycle length must be at least {MIN_CYCLE_LENGTH}, but was {length} for cycle {cycle_name}"
+        ))
+    }
```

Comment:
let's prefix these with `periodic_` for (e.g. `periodic_cycle_length_not_power_of_two`)